src/Syntax/Resolve.hs

```diff
@@ -293,11 +298,21 @@ reExpr r@(Ascription e t a) = Ascription
                           <$> reExpr e
                           <*> reType r t
                           <*> pure a
-reExpr (Record fs a) = Record <$> traverse (traverse reExpr) fs <*> pure a
-reExpr (RecordExt e fs a) = RecordExt
-                        <$> reExpr e
-                        <*> traverse (traverse reExpr) fs
-                        <*> pure a
+reExpr e@(Record fs a) = do
+  let ls = map fst fs
+      nub = Set.toList . Set.fromList
+  case ls \\ nub ls of
+    l:_ -> throwError $ NonLinearRecord e l
+    _ -> pure ()
+  Record <$> traverse (traverse reExpr) fs <*> pure a
+reExpr (RecordExt e fs a) = do
+  let ls = map fst fs
+      nub = Set.toList . Set.fromList
+  case ls \\ nub ls of
+    l:_ -> throwError $ NonLinearRecord e l
```

Comment:
Should this cover the whole expression, or is it OK just to include the extended part?